fix 3213: athena integration column 'year' cannot be resolved

spa-raj opened the pull request:
## What does this PR change?
*  This PR fixes the athena integration query logic for fetching Cloud Usage Reports (CUR) by introducing a change to date filtering logic.

Updated the `GetPartitionWhere` method to use `date_format` for extracting year and month from the `line_item_usage_start_date` field, ensuring compatibility with the dataset's date format.

## Does this PR relate to any other PRs?
* No

## How will this PR impact users?
* Users will be able to see their AWS cloud usage reports in the opencost dashboard ui

## Does this PR address any GitHub or Zendesk issues?
* Closes #3213 

## How was this PR tested?
* This PR was tested 
 i. Deploying needed resources per AWS integration guide
 ii. Connecting Athena to CUR output bucket via Glue crawler
 iii. Deploy OpenCost using the Tilt dev tooling.
 iv. Observing Athena query logs for query success.
 v. Verifying the AWS CUR report successfully populating in the opencost dashboard ui.
 
 Dashboard screenshot
<img width="2534" height="1341" alt="image" src="https://github.com/user-attachments/assets/1ece0af7-9712-4227-83d1-ff8232a9bbe9" />

Tilt costmodel-tests and athena success screenshot
<img width="2548" height="1350" alt="image" src="https://github.com/user-attachments/assets/ae3ab54b-2371-414f-8b4f-9ce727119e16" />
<img width="2548" height="941" alt="image" src="https://github.com/user-attachments/assets/44a158f7-d0ed-4b36-bd19-f4a5b7a49945" />


## Does this PR require changes to documentation?
* No

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* It should be part of next release but I don't have permissions to label it.

spa-raj committed:
fix 3213: athena integration column 'year' cannot be resolved

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

ameijer commented on pkg/cloud/aws/athenaintegration.go:
```diff
@@ -329,7 +329,8 @@ func (ai *AthenaIntegration) GetPartitionWhere(start, end time.Time) string {
 	for !month.After(endMonth) {
-		disjuncts = append(disjuncts, fmt.Sprintf("(year = '%d' AND month = '%d')", month.Year(), month.Month()))
+		disjuncts = append(disjuncts, fmt.Sprintf("(date_format(line_item_usage_start_date, '%%Y') = '%d' AND date_format(line_item_usage_start_date, '%%m') = '%02d')",
+			month.Year(), month.Month()))
```
let's freeze a 1.0 version, and make a 2.0 version. can make into an athena config. field added to show what type

spa-raj committed:
Merge branch 'opencost:develop' into fix-3213-COLUMN_NOT_FOUND-line-311-Column-'year'-cannot-be-resolved

spa-raj committed:
fix: support CUR 1.0 and 2.0 partition formats in Athena integration
Fixes #3213 by adding CURVersion configuration to handle different partitioning schemes between AWS CUR versions:
  - CUR 1.0 uses 'year' and 'month' columns for partitioning
  - CUR 2.0 uses date_format functions on line_item_usage_start_date

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Updated tests to include CUR version in Athena config

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
feat: add intelligent CUR 2.0 billing_period partition detection for AWS Athena

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj commented:
# Major Enhancement: Intelligent CUR 2.0 Partition Detection

 I've significantly enhanced the AWS Athena integration based on extensive research and testing. The latest commits introduce intelligent partition detection that automatically adapts to different CUR configurations.

## Latest Changes

### Dynamic Partition Detection

  - New Method: HasBillingPeriodPartitions() - Dynamically detects if CUR 2.0 tables have billing_period partitions via SHOW PARTITIONS
  - Smart Query Optimization: Automatically uses the most efficient partition format based on actual table structure
  - Enhanced Compatibility: Works with all CUR setups, including edge cases where CUR 2.0 tables lack billing_period partitions

### Improved Partition Strategy

  ```
// Now dynamically chooses the best approach:
  if ai.CURVersion == "1.0" {
      // Uses year/month columns
      "(year = '2024' AND month = '8')"
  } else if hasBillingPeriodPartitions {
      // Uses efficient billing_period partitions  
      "(billing_period = '2024-08')"
  } else {
      // Fallback for tables without billing_period partitions
      "(date_format(line_item_usage_start_date, '%Y') = '2024' AND date_format(line_item_usage_start_date, '%m') = '08')"
  }
```

### Comprehensive Testing

  - **Tested with real data on both CUR 1.0 and CUR 2.0 implementation**
  - Mock Implementation: Added mockAthenaQuerier and mockAthenaIntegration to prevent real AWS API calls during unit tests
  - Complete Test Coverage: All CUR 1.0/2.0 scenarios, cross-year boundaries, fallback cases, and edge conditions
  - Zero Test Failures: All AWS integration tests now pass reliably

#### Technical Deep Dive

  My research revealed that in CUR 2.0:
  - billing_period is a partition metadata column (not a data column) - explains why it doesn't appear in information_schema.columns
  - Some CUR 2.0 setups may not have billing_period partitions depending on configuration
  - Partition detection via SHOW PARTITIONS is the reliable way to determine availability
  - Query performance is significantly better when using partition columns vs. date functions

#### Impact

  - Fixes #3213: Resolves the original "Column 'year' cannot be resolved" error
  - Performance: Uses most efficient querying approach for each CUR configuration
  - Reliability: Robust fallback ensures compatibility across all AWS setups
  - Maintainability: Comprehensive test coverage prevents regressions

#### Ready for Review

The implementation now handles all CUR scenarios intelligently and has been thoroughly tested. This should provide a robust solution for AWS cost data ingestion across different CUR configurations.

spa-raj committed:
Merge branch 'opencost:develop' into fix-3213-COLUMN_NOT_FOUND-line-311-Column-'year'-cannot-be-resolved

spa-raj committed:
Merge branch 'opencost:develop' into fix-3213-COLUMN_NOT_FOUND-line-311-Column-'year'-cannot-be-resolved

ameijer commented on pkg/cloud/aws/athenaconfiguration.go:
```diff
@@ -190,6 +201,18 @@ func (ac *AthenaConfiguration) UnmarshalJSON(b []byte) error {
+	} else {
+		// Default to 2.0 if not specified
+		ac.CURVersion = "2.0"
+	}
```
@Sean-Holcomb here's the switch statement that we discussed adding last time, think we are good to merge?

Sean-Holcomb replied:
We should default to 1.0, to not break existing users

spa-raj committed:
Merge branch 'opencost:develop' into fix-3213-COLUMN_NOT_FOUND-line-311-Column-'year'-cannot-be-resolved

Sean-Holcomb reviewed (approved):
Thanks for all the tests. Looks good

spa-raj committed:
Merge branch 'opencost:develop' into fix-3213-COLUMN_NOT_FOUND-line-311-Column-'year'-cannot-be-resolved

spa-raj commented:
@ameijer 2 workflows are pending approval. I guess those are needed for merging.
